File: app/src/main/java/dev/pitlor/smssync/datasources/daos/MessageDao.java
package dev.pitlor.smssync.datasources.daos;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;

import java.time.OffsetDateTime;
import java.util.List;

import dev.pitlor.smssync.datasources.dto.MessageDTO;
import dev.pitlor.smssync.datasources.entities.Message;

@Dao
public interface MessageDao {
    @Transaction
    @Query("SELECT * FROM message")
    LiveData<List<MessageDTO>> getAll();

    @Transaction
    @Query("SELECT * FROM message WHERE id IN (:messageIds)")
    LiveData<List<MessageDTO>> loadAllByIds(int[] messageIds);

    @Query("SELECT SUM(id) FROM message GROUP BY id")
    LiveData<Integer> size();

    @Query("SELECT date FROM message ORDER BY date DESC LIMIT 1")
    LiveData<OffsetDateTime> timeOfLastSavedText();

    @Insert
    void insertAll(List<Message> messages);

    @Delete
    void delete(Message message);
}

File: app/src/main/java/dev/pitlor/smssync/datasources/daos/MessageDao.kt
package dev.pitlor.smssync.datasources.daos

import androidx.lifecycle.LiveData
import androidx.room.*
import dev.pitlor.smssync.datasources.Message
import dev.pitlor.smssync.datasources.MessageWithContact
import java.time.OffsetDateTime

@Dao
interface MessageDao: BaseDao<Message> {
    @Query("SELECT SUM(id) FROM message GROUP BY id")
    fun getSize(): LiveData<Int>

    @Query("SELECT date FROM message ORDER BY date DESC LIMIT 1")
    fun getTimeOfLastSavedText(): OffsetDateTime?

    @Transaction
    @Query("SELECT * FROM Message GROUP BY threadId ORDER BY MAX(date)")
    fun getAllThreads(): LiveData<List<MessageWithContact>>
}